core/version/gear.ex

```diff
@@ -113,63 +115,92 @@ defmodule AntikytheraCore.Version.Gear do
   @typep gear_dependencies :: MapSet.t()
   @typep gear_and_deps_pair :: {GearName.t(), gear_dependencies}
 
-  defun install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok do
+  defun install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok | :error do
     if Antikythera.Env.running_with_release?() do
       do_install_gears_at_startup(gear_names)
     else
       :ok
     end
   end
 
-  defunp do_install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok do
+  defunpt do_install_gears_at_startup(gear_names :: v[[GearName.t()]]) :: :ok | :error do
+    # Tests need `__MODULE__.` to mock these functions.
     gear_and_deps_pairs =
-      Enum.map(gear_names, fn g -> {g, gear_dependencies_from_app_file(g, gear_names)} end)
-
-    pairs_not_installed =
-      install_gears_whose_deps_met(gear_and_deps_pairs, MapSet.new(), fn gear_name ->
-        try do
-          install_or_upgrade_to_next_version(gear_name)
-        rescue
-          e -> L.error("Failed to install #{gear_name}: #{Exception.message(e)}")
-        end
+      Enum.map(gear_names, fn g ->
+        {g, __MODULE__.gear_dependencies_from_app_file(g, gear_names)}
       end)
 
+    {pairs_not_installed, num_failed_install} =
+      __MODULE__.install_gears_whose_deps_met(
+        gear_and_deps_pairs,
+        MapSet.new(),
+        0,
+        fn gear_name ->
+          try do
+            install_or_upgrade_to_next_version(gear_name)
+          rescue
+            e ->
+              L.error("Failed to install #{gear_name}: #{Exception.message(e)}")
+              :error
+          end
+        end
+      )
+
     Enum.each(pairs_not_installed, fn {gear_name, deps} ->
       L.error("#{gear_name} is not installed due to unmatched dependencies: #{inspect(deps)}")
     end)
+
+    num_installable = length(gear_names)
+    num_installed = num_installable - length(pairs_not_installed) - num_failed_install
+    # version_upgrade_test starts Antikythera without gears.
+    # Then, we can't check number of installed gears.
+    needs_check = Antikythera.Env.runtime_env() != :local
```

Comment:
`is_enough_gears_installed` seems to handle even if the gear is empty. Isn't `needs_check` unnecessary?

Reply:
Sorry if I misunderstand your question.

If `num_installable` is 0, `is_enough_gears_installed` becomes `false` and this returns `:error`.
Since `install_gears/0` matches this return value with `:ok`, Antikythera initialization will throw MatchError.
Since Antiktyhera isn't initialized, `:rpc.call/4` in `Mix.Tasks.AntikytheraLocal.VersionUpgradeTest`, which run on `ANTIKYTHERA_COMPILE_ENV=local` , will throw an exception.
 
`needs_check` is needed to make this function return `:ok` even if `num_installable == 0`.

Reply:
Thaks. I read the code incorrectly.

Reply:
@Fumipo-Theta
I have a question about the internal environment. Is there possibility that there are no gears when starting Antikythera?

Reply:
In the internal environment, Erlang nodes start after the artifacts of two gears are generated and Antikythera release is made.
So, I think there should be at least two gears when starting Antikythera.
I made sure Antikythera starts with the two gears.

Reply:
Thank you for the description in public place.

Note: Internal Slack discussion about it:
https://accessjp.slack.com/archives/CDEN4MANQ/p1605694056455100